### PR TITLE
Platform-specific file locking for kubeconfig

### PR DESCRIPTION
**Pull request by rosspeoples**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Previously when writing the kubeconfig file, it would write out  a `.lock` file which was used to prevent concurrent writes to the file. This PR replaces that mechanism with a more robust platform-specific locking mechanism using `LockFileEx` for Windows or `flock` for Linux.

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:
This is my first pass and I made the assumption that we didn't want to change the signature of `WriteToFile` which is why it's structured this way. ~~Also, while previously we were pre-locking all the files at once, they are now locked on-demand.~~ This maintains the pre-locking behavior.

#### Does this PR introduce a user-facing change?

```release-note
This updates the etcd dependency to 3.5.1 and is the first step to transitioning to platform-native file locking for handling concurrent writes to kubeconfig files.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
Related to conversation found here: https://github.com/kubernetes/kubernetes/pull/92513

```docs

```

### CODE REVIEW

**Comment by jpbetz** [comment]
/lgtm

This file lock code is critical to etcd and is pretty well thought through.

I don't have an opinion on if we should be using the etcd libraries directly (do we already use them elsewhere?),

**Comment by BenTheElder** [comment]
> Previously when writing the kubeconfig file, it would write out a .lock file which was used to prevent concurrent writes to the file. This PR replaces that mechanism with a more robust platform-specific locking mechanism using LockFileEx for Windows or flock for Linux.

How will this behave when another process uses the current implementation concurrently with this new approach?

> Also, while previously we were pre-locking all the files at once, they are now locked on-demand.

I'm not sure this change is safe, IIRC they were all pre-locked so the contents can be read and then written back without race condition.

**Comment by deejross** [comment]
I knew we were using it in k/k: https://github.com/kubernetes/kubernetes/blob/master/go.mod#L80 so I figured it was pretty safe to include in client-go.

**Comment by deejross** [comment]
> IIRC they were all pre-locked so the contents can be read and then written back without race condition

That's a good point, I'll see if I can cache the locks to allow for the pre-locking to remain.

**Comment by BenTheElder** [comment]
> How will this behave when another process uses the current implementation concurrently with this new approach?

More specifically: If I have binary A using client-go from kubernetes 1.23 and I have binary B using client-go from kubernetes 1.24 with this change, and they both attempt to lock the configs, one using the .lock file and the other with this approach, presumably with e.g. flock on linux both processes will "lock" the file with their respective approaches and race reading/writing the configs?

Repeat question for each platform's unique approach in the new implementation.

**Comment by deejross** [comment]
> If I have binary A using client-go from kubernetes 1.23 and I have binary B using client-go from kubernetes 1.24 with this change, and they both attempt to lock the configs, one using the .lock file and the other with this approach, presumably with e.g. flock on linux both processes will "lock" the file with their respective approaches and race reading/writing the configs?

Another great point, I guess it would make sense to deprecate the `.lock` files, but continue using both approaches for a few releases. Priority would be to check for platform-specific locks first, then checking for `.lock` files in order to allow for a smoother migration in the future.

**Comment by BenTheElder** [comment]
I think that sounds right, there probably needs to be some decision about how many releases. Not sure on that part ... 🤔 

Also, we might need to check on client libraries that aren't client-go and this probably will need a release note to warn that other implementations need to be updated to match? (right now this PR has release-note=NONE).

**Comment by deejross** [comment]
@BenTheElder any thoughts on the latest changes?

File: staging/src/k8s.io/client-go/tools/clientcmd/loader.go [hunk 1]
```diff
@@ -26,6 +27,7 @@ import (
 	"strings"
 
 	"github.com/imdario/mergo"
+	"go.etcd.io/etcd/client/pkg/v3/fileutil"
```
Comment (dims) [inline staging/src/k8s.io/client-go/tools/clientcmd/loader.go:30]: not too happy about this ... our client-go depending on "go.etcd.io/etcd/client/pkg/v3/fileutil"

cc @liggitt

**Comment by liggitt** [comment]
> not too happy about this ... our client-go depending on "go.etcd.io/etcd/client/pkg/v3/fileutil"

agreed. I actually think the lock file behavior is generally problematic and we should figure out how to get rid of it (xref https://github.com/kubernetes/kubernetes/issues/32606, https://github.com/kubernetes/kubernetes/issues/67676)

**Comment by liggitt** [comment]
were lock-free alternatives like writing to a tmp file and moving to replace considered?

**Comment by liggitt** [comment]
> The current mechanism uses `.lock` files, but they are not completely atomic.

right, and they have other problems like requiring the entire directory to be writeable instead of just the kubeconfig file

File: staging/src/k8s.io/client-go/tools/clientcmd/loader.go (hunk 1 above)
Comment (deejross) [reply staging/src/k8s.io/client-go/tools/clientcmd/loader.go:30]: It's already in other areas of the codebase: https://github.com/kubernetes/kubernetes/blob/master/go.mod#L80

Comment (liggitt) [reply staging/src/k8s.io/client-go/tools/clientcmd/loader.go:30]: that's not really relevant... adding dependencies to k8s.io/{api,apimachinery,client-go} pushes them onto far more consumers

**Comment by BenTheElder** [comment]
sorry, I think these changes address my initial comments, but the other concerns mentioned seem valid.

I don't know if there's a good path forward on this. I do think tools are going to race merging changes without some sort of lock (e.g. in my case users concurrently creating [KIND](https://kind.sigs.k8s.io) clusters all writing to the default $KUBECONFIG), but ...

> were lock-free alternatives like writing to a tmp file and moving to replace considered?

Atomic file rename would solve that readers unaware of the lock may try to read during writing, but the locks are also important for the problem of "two processes are reading, mutating, then writing back config", e.g. `gcloud container clusters get-credentials`, `kind create cluster`, etc. -- in that case we need to keep the files locked until the read-merge-write cycle is complete, which is something client-go supports today, and atomic rename alone doesn't address.

The approach currently in this PR would still support that and address the "directory needs to be writeable" concern once we remove the old .lock code after all these tools are updated to match after some grace period of supporting both modes.

IMO writing by atomic rename would be great, but it would bring back the "directory must be writeable" problem. 

You can't just write to a temp directory when doing atomic rename because we can only do a `rename(2)` on the same filesystem/disk, and we can't guarantee that TMPDIR is on the same disk as $KUBECONFIG. Usually you would write to a temp file adjacent to the target path (by appending some random suffix) and then rename, which will still require the directory to be writeable.

**Comment by deejross** [comment]
@BenTheElder what would be the next steps on this?

**Comment by BenTheElder** [comment]
: thinking:

I don't know if this change needs anything so heavyweight as a KEP, but you might want to at least bring it up with SIG CLI in some way and/or api-machinery (client-go).

https://github.com/kubernetes/kubernetes/pull/107354#discussion_r788794671 seems like the main blocker in review so far, client-go dependencies can be something of a headache, perhaps the subset we need can be re-implemented without additional dependencies (or forked into one of our existing dependencies?), but before going on that route I'd probably try to make sure OWNERS agree.

I only **own** this from a dependency review standpoint, I just also happen to have some familiarity with this code and problem ...
https://github.com/kubernetes/community/blob/4026287dc3a2d16762353b62ca2fe4b80682960a/contributors/devel/sig-architecture/godep.md#reviewing-and-approving-dependency-changes

**Comment by kensipe** [comment]
Looks like I'm late to the game here.   For awareness, there has been a [long standing request](https://github.com/golang/go/issues/33974) to add this feature in Golang... technically it already exists in Go but lives under `internal` and thus is not exposed.   There is a [design proposal](https://go.googlesource.com/proposal/+/refs/changes/77/201277/18/design/33974-add-public-lockedfile-pkg.md) which merged Mar 20, 2020 however it has not be "Accepted".  As noted on the issue, Ian is pushing for "faster consideration", which I don't entirely know what that means but trust him.
Prior to fully reading the proposal, I created an example of what I thought should be exposed from the Go core:  https://github.com/kensipe/flock  Now that I more fully understand the proposal.. I'm planning on refactoring that project.  The intention of my project is to match the proposal close enough to not need any changes (other than imports) to adjust to the addition of it to the golang core. (which we should expect to be months away).   The solution that exists there today is a lower level lock, rlock, unlock functional.  That is not the level the Go team wants to expose... the expectation is to expose several files from https://github.com/golang/go/tree/f229e70/src/cmd/go/internal/lockedfile. in particular `[lockedfile.go](https://github.com/golang/go/blob/f229e7031a6efb2f23241b5da000c3b3203081d6/src/cmd/go/internal/lockedfile/lockedfile.go) and [mutex.go](https://github.com/golang/go/blob/f229e7031a6efb2f23241b5da000c3b3203081d6/src/cmd/go/internal/lockedfile/mutex.go)

IMO if we are going to properly support this, we should align with coming expectations.

**Comment by deejross** [comment]
Closing this due to the general consensus being we don't to bring a new dependency into client-go, and would prefer to utilize locking functionality exposed from the Go standard library.

**Pull request closed without merging**